### ISSUE TITLE
Ollama

### ISSUE THREAD

**Issue by euroblaze** [euroblaze/voltAIc#2]

**Comment by CaoHoangTung** [comment]
@euroblaze 

## How to leverage the config of the Ollama server?

### Modelfile

Ollama’s models are inspired by Docker image. The idea is to package everything needed to get a model up and running, with further configuration, into 1 reproducible file. 

Ollama can then pick up this file and setup a model image, which can be loaded and used for inference. The model image can also be uploaded into Ollama hub, allowing others to reuse our model & setup

A few features of Modelfile:

- Extends and build on top of existing model. You can finetune a public model like llama or mistral
- [Embed a system prompt or preparatory conversation for the model](https://github.com/ollama/ollama/blob/main/docs/modelfile.md#system)
- Setup Prompt Template
- [Set various inference parameter](https://github.com/ollama/ollama/blob/main/docs/modelfile.md#parameter)
    - Perplexity control (tune coherent and diversity)
    - Sampling method
    - Output space search parameter (top_k probability)
    - Threads and core used for inference

### GGUF

Sometimes a public model is not really fine-tuned to a given task, even with extensive prompt engineering. There might be different problems, with accuracy & speeds.

If you are able to finetune a model on a specific dataset or quantize it for speed improvement, we can export the model into a GGUF file, which can be imported and run for inference on Ollama

### Model storage:

Ollama store models in layers, another idea taken from Docker. This allows saving multiple model version more effectively. Each model, defined by a tag (e.g. `mistral:7b`, `gemma:2b`) consist of a stack of layer, with each one record a segment of model state/configuration. For example, given the following Modelfile:

```
FROM mistral:7b

SYSTEM "You are an assistant specialized in answering questions about dolphin"
```

If you build a model from this Modelfile, it will consist of:

- All of the layers from `mistral:7b`
- A layer record the change of model state after receiving the system prompt

Loading a model is as simple as applying a model state changes one by one and load it in memory for inference.

This way, if you need to build 2 different task-specific model based on `mistral:7b`, the computer will only store the base layers once

You can specify a directory to store the layers, using the `OLLAMA_MODELS` environment variable

### Model loading & inference

Ollama will load models on demand, so one Ollama server can serve a multitude of models (limited by the hardware resource). Upon the first request for inference of a given model, Ollama will attempt to load it in to memory, optionally GPU, and try to do inference on it. Some settings allow developer to set the timeout for model to stay in memory. After this timeout, Ollama will unload the model. This will be very convenient during development of model, where you can try different model version on one single computer.

However, it is recommended that, in production, we only load and infer on one model only per Ollama server. This is to ensure the model is loaded into the correct device and run at maximum performance.

## How to update the Docker without loss of config/settings and log files?

This questions concerns 2 different things. The part about log will be explained in the next questions. For this question, we address config/settings persistence.

### Ollama persistence:

The Ollama server application is highly stateless. We only need to specify where on the filesystem to lookup model layers to startup a Ollama server application. 

The runtime config is defined by the consumer code (your application code). For example:

- What model to run
- How long to keep it in memory

The only thing that need to be persisted is the model layer directory. For Docker, this is facilitated by a [Docker volume](https://docs.docker.com/storage/volumes/)

## How to store and read log files on host, not container.
Since ollama log is in stdout/stderr in container [reference](https://github.com/ollama/ollama/blob/main/docs/troubleshooting.md), we can custom build a dockerfile from ollama base image, redirect the log to a file using `CMD`. Then using Docker volume, we can mount the volume from that file to host.

## How is the Ollama server networked via micro-services to other services?

- For a docker based deployment, we make use of [Docker network](https://docs.docker.com/network/) to connect the Ollama server to the other services in our application
- Two services can only reach each other if connected into at least one common network
- The services are reachable by their name.

For example: given those two services defined using a Docker Compose template

```yaml
services:
  voltaic-app:
    image: euroblaze/voltaic:latest
    networks:
	    app_nw:
  ollama:
    image: ollama/ollama:0.1.9
    volumes:
      - ./ollama:/root/.ollama
    networks:
	    app_nw:
	  
networks:
  app_nw:

```

The `voltaic-app` service can reach `ollama` service through `http://ollama:11434`

**Comment by euroblaze** [comment]
@CaoHoangTung 

> If you are able to finetune a model on a specific dataset or quantize it for speed improvement, we can export the model into a GGUF file, which can be imported and run for inference on Ollama

The trained models from llama-factory can be saved to GGUF format?

**Comment by CaoHoangTung** [comment]
> The trained models from llama-factory can be saved to GGUF format?

Have been discussed in https://github.com/euroblaze/voltAIc/issues/10#issuecomment-2052068620

**Comment by CaoHoangTung** [comment]
@euroblaze 
# Upgrading ollama container to a newer version

Here's the command to upgrade the Ollama version, with a few seconds downtime for tearing down the old container and bringing up a new one

```bash
sudo docker pull ollama/ollama:latest \
&& sudo docker stop ollama \
&& sudo docker rm ollama \
&& sudo docker run -d -v ollama:/root/.ollama -p 11434:11434 --name ollama ollama/ollama
```
